### ISSUE TITLE
Review Hotfix/Patch Process

### ISSUE THREAD

**Issue by lewri** [CorsixTH/CorsixTH#2639]
https://github.com/CorsixTH/CorsixTH/wiki/Hotfixes

While I wrote the Hotfix process in 2021 hastily for `0.65.1` it likely wasn't a well balanced method.

Ideally when a Hotfix is pushed out there should be a savegame bump. In `0.66` we assumed both `0.65` and `0.65.1` were the same version, but they are not.

The game's releases ideally needs a set differentiator at all times. The trouble is, merely bumping the version may cause some odd issues for the master branch. E.g. `0.68.0` releases as savegame version 194. During development we find out that something needs to be patched in release, however master is already at 197. If `0.68.1` logically gets assigned version 195; would all savegame version increments in master now need to be increased by 1? As such, master becomes 198.

In short, I'm not sure if the current process is fit for purpose. What alternatives are there?

**Comment by lewri** [comment]
Another thought is if we don't want to interfere with bumping master on a hotfix, is release a savegame bump as e.g. `194.1` -- we usually deal with integers so we'd have to be prepared to allow for incremented numbers.

**Comment by Alberth289346** [comment]
If you apply all missing patches from development to the release, it's going to work. Whether you can still call that a "release" is questionable then.

----
A second option is to never patch a release. It does what it does. If it doesn't work for some feature, it's a known defect, possibly with a work-around for the player.

A way to fix something here is to make a new release from development.

----
A third option is to include all development patches in the release, but hide the differences that they cause. This doesn't look very feasible to me.

----
A fourth option is to allow diverging patch streams, that is, give up ever merging them again.
That works. You patch the release stream separately. It makes games created with a patched release non-compatible with anything else though.  In other words, once you start a game in a patched release, you must stay in that release stream for that game, and may only upgrade for more patches in the release.

----
Then the fifth option, do some kind of merge of the divergent streams afterwards.

I mostly consider all methods to try merging diverged patch streams again as non-feasible, at least if you don't want to create a lot of work for yourself and take a lot of risks.  
But for the sake of discussion, let's assume we really want to do this.

In your example with a release at 194 and development at 197, let's assume we patch the release to some new number `X` (I don't care what `X` should be at this point.)
Obviously, the development stream should also be fixed, let's use 198 for that.

As soon as you add `X`, the release is not compatible with development anymore. Games started in a release cannot be used in development versions. You already noted that.

So we have:
- a release with patches 194 and `X`, and
- a development stream with patches 194, 195, 196, 197, and 198.

How are we going to make them compatible again?
I think it needs a patch with the merged changes of `-X + 195 + 196 +197 + 198`. In other words, you need not only write patch `X`, but also the reverse patch.

Now if all patches are independent, this is weird but doable. If they are **not** independent (eg 196 interferes with 198), I don't know but it's going to be messy I think.

----
The usual solution to patching releases and still allow upgrade, is by not changing the data that is saved.
We currently don't have that. (We basically create a new major version with every patch, and we then make it work by extending `afterLoad`.)

If the saved data is not changed for patched releases, upgrade is trivial and painless.

So that's another path. Create better control of how and what gets saved in a game.
A much longer path, but it should eventually win from all other options here, I think.

**Comment by lewri** [comment]
Currently the way the hotfix is worded is to ship only the code to rectify the scenario and nothing more. Would definitely prefer seeing to keep that notion. I'm not entirely sure which option above might fit that.

**Comment by Alberth289346** [comment]
> Currently the way the hotfix is worded is to ship only the code to rectify the scenario and nothing more.

Reduced scope is probably a good idea, but as far as I can see, pretty much any code change will cause a change in the data of the game, and thus in the saved data. In other words, I think as soon as you save a game, that file is not compatible with development anymore.

Disabling the hotfix at a later stage makes no difference at all for such saves, since loading the file directly rebuilds the data as it was at the time of save in memory, including the changes caused by the hotfix. Not applying the hotfix doesn't revert the hotfix changes. The loaded file is therefore still not compatible.

You may be lucky in the sense that the differences are not causing trouble (and of course the scope of the changes have an influence on it). That may make them look like they are  compatible, but I think it is simply taking the risk rather than being sure it's ok.

**Comment by TheCycoONE** [comment]
Could always jump by 10 or 100 post release so there is room too. It has never been an issue though.

**Comment by lewri** [comment]
> Could always jump by 10 or 100 post release so there is room too. It has never been an issue though.

This likely would also not cause any issues once I get round to #2527 / #2518. 10 is probably safe enough giving how sparingly we would actually use it.

**Comment by Alberth289346** [comment]
Say we skip the first 10 as reserved for the release.

In the example, release is at 194, so 195 becomes the patch for the release.
Development starts at 204 with the first patch (195 in the example), and then 205 and 206 (for 196 and 197 in the example) 207 is then the development variant of release patch 195.

Now when a user loads a save of the patched release in master, shouldn't it skip 207 ?

**Comment by TheCycoONE** [comment]
Yes, when you merge back you'll need to gate the change or make it repeatable then bump the save version again and redo the afterload criteria. It only really saves distinguishing the two.

**Comment by lewri** [comment]
Solutions are most likely easily manageable.
If a hotfix needs an afterLoad you can apply it to development and a branch with different savegame numbers. How you handle the latter development version for the hotfixed ones could do:
```lua
-- devel patch was 207; hotfix used 195; next release began at 204
if old < 207 and not (old >= 195 and old < 204) then
  ...
end
```
Though that could get complicated fast. It may be safer to still use a hotfix marker.
```lua
-- devel patch was 207; hotfix used 195
if old < 207 and not TheApp.map.hotfix195 then
  ...
end
```
^ It doesn't have to be stored in `map`, but it can't be `app`. I'm less against not cleaning up hotfix markers if they get labelled by the savegame version; when you move to another level or restart the level they will clear anyway.

If it doesn't need an afterLoad, then it doesn't really matter. You still bump both branches but development can remain without a savegame bump. Hotfix will need a bump though just to declare version.

> Now when a user loads a save of the patched release in master, shouldn't it skip 207 ?

Master builds should likely always be considered like a `nightly` in my opinion, and shouldn't be treated as being stable or have guaranteed compatibility.

**Comment by TheCycoONE** [comment]
An ordered patch table could have been used the whole time. It's not a terrible solution, i've seen it in the industry.

**Comment by Alberth289346** [comment]
> Master builds should likely always be considered like a nightly in my opinion, and shouldn't be treated as being stable or have guaranteed compatibility.

Sure, but if you get a savegame from a release, it's quite useful if you can run it in master too.

**Comment by ARGAMX** [comment]
Essentially, to close this task, we just need to write in the Wiki that we’re incrementing `SAVEGAME_VERSION` by 10 after the release? And that’s it?
